### PR TITLE
Make safely_acquire_lock_for_table aware of schemas / partitions

### PR DESCRIPTION
**Pull request by rkrage**
As described [here](https://github.com/braintree/pg_ha_migrations/pull/85#discussion_r1317471168), there are a few issues with the locking implementation if a fully qualified table name is provided and/or the table is partitioned. This aims to fix those things and additionally adds some guards to detect nested locks (this is generally bad and can lead to deadlocks).

Also included is a refactor to use structs whenever possible to avoid complicated nested array structures

### CODE REVIEW

File: spec/safe_statements_spec.rb [hunk 1]
```diff
@@ -3484,6 +3496,95 @@ def up
...
             end
 
+            it "does not wait to acquire a lock if same table in different schema is blocked" do
+              stub_const("PgHaMigrations::LOCK_TIMEOUT_SECONDS", 1)
+
+              ActiveRecord::Base.connection.execute("CREATE TABLE partman.bogus_table(pk SERIAL, i INTEGER)")
+
+              begin
+                thread = Thread.new do
+                  migration.safely_acquire_lock_for_table("partman.bogus_table") { sleep 2 }
```
Comment (jcoleman) [inline spec/safe_statements_spec.rb:3506]: I wonder if we should do an explicit lock here rather than relying on our code (since our own code is also what's under test).

File: spec/safe_statements_spec.rb [hunk 2]
```diff
@@ -3484,6 +3496,95 @@ def up
...
+                sleep 1.1
+
+                expect(PgHaMigrations::BlockingDatabaseTransactions).to receive(:find_blocking_transactions)
+                  .at_least(3)
+                  .times
+                  .and_call_original
+
+                migration.suppress_messages do
+                  migration.safely_acquire_lock_for_table(table_name) do
+                    expect(locks_for_table(table_name, connection: alternate_connection)).not_to be_empty
```
Comment (jcoleman) [inline spec/safe_statements_spec.rb:3546]: Should we also expect that there are no explicit locks (not sure if we distinguish) on `bogus_table_default`? If we don't have a way to distinguish I suppose we could expect that `find_blocking_transactions` returns empty before this happens?

File: lib/pg_ha_migrations/safe_statements.rb [hunk 3]
```diff
@@ -553,6 +542,31 @@ def _partitioned_table?(schema, table)
...
+    SQL
+
+    if include_sub_partitions
+      sub_partitions = schemas_and_tables.each_with_object([]) do |(child_schema, child_table), arr|
+        if _partitioned_table?(child_schema, child_table)
+          arr.concat(_partitions_for_table(child_schema, child_table, include_sub_partitions: true))
+        end
+      end
+
+      schemas_and_tables.concat(sub_partitions)
```
[the remark this answers is not in the document]
Comment (rkrage) [reply lib/pg_ha_migrations/safe_statements.rb]: Addressed in https://github.com/braintree/pg_ha_migrations/pull/87/commits/bd2c1e92e2cfd7765f579f6baec38bf66f4a1a22

File: spec/safe_statements_spec.rb (hunk 1 above)
Comment (rkrage) [reply spec/safe_statements_spec.rb]: Addressed in https://github.com/braintree/pg_ha_migrations/pull/87/commits/bd2c1e92e2cfd7765f579f6baec38bf66f4a1a22

File: spec/safe_statements_spec.rb (hunk 2 above)
Comment (rkrage) [reply spec/safe_statements_spec.rb]: Addressed in https://github.com/braintree/pg_ha_migrations/pull/87/commits/bd2c1e92e2cfd7765f579f6baec38bf66f4a1a22

File: lib/pg_ha_migrations/safe_statements.rb [hunk 4]
```diff
@@ -587,16 +601,38 @@ def exec_migration(conn, direction)
   end
 
   def safely_acquire_lock_for_table(table, &block)
+    nested_lock = Thread.current[__method__].present?
```
[the remark this answers is not in the document]
Comment (rkrage) [reply lib/pg_ha_migrations/safe_statements.rb]: Addressed in https://github.com/braintree/pg_ha_migrations/pull/87/commits/bd2c1e92e2cfd7765f579f6baec38bf66f4a1a22

File: lib/pg_ha_migrations/safe_statements.rb [hunk 5]
```diff
@@ -553,6 +542,31 @@ def _partitioned_table?(schema, table)
     SQL
   end
 
+  def _partitions_for_table(schema, table, include_sub_partitions: false)
```
[the remark this answers is not in the document]
Comment (rkrage) [reply lib/pg_ha_migrations/safe_statements.rb]: So I'm going back and forth on this one. With the most recent changes, this method accepts a table struct and returns an array of table structs. I guess we could have two methods `_partition_structs_for_table` (or something similar) for use internally, and `partitions_for_table` that would accept a string and return a list of fully qualified child tables.

File: lib/pg_ha_migrations/struct.rb [hunk 6]
```diff
@@ -0,0 +1,59 @@
+module PgHaMigrations::Struct
+  Table = Struct.new(:name, :schema) do
+    delegate :inspect, to: :name
+
+    def fully_qualified_name
+      @fully_qualified_name ||= [
+        ActiveRecord::Base.connection.quote_schema_name(schema),
+        ActiveRecord::Base.connection.quote_table_name(name),
+      ].join(".")
+    end
```
Comment (rkrage) [inline lib/pg_ha_migrations/struct.rb:10]: We _could_ add even more methods here, like `partitioned?` and `valid?` (for the index struct below), but I like that these are data objects that only contain light validation and formatting

File: lib/pg_ha_migrations/safe_statements.rb [hunk 7]
```diff
@@ -490,14 +482,16 @@ def _quoted_partman_schema
...
+
+    [table.schema, table.name].each do |identifier|
       if identifier.to_s !~ /^[a-z_][a-z_\d]*$/
         raise PgHaMigrations::InvalidMigrationError, "Partman requires schema / table names to be lowercase with underscores"
       end
     end.join(".")
   end
 
-  def _schema_and_table_for(table)
+  def _struct_from_table_name(table)
```
Comment (jcoleman) [inline lib/pg_ha_migrations/safe_statements.rb:494]: I think we should move this code directly into `Struct::Table.from_table_name`.

File: lib/pg_ha_migrations/struct.rb [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+    def present?
+      name.present? && schema.present?
+    end
+  end
+
+  Index = Struct.new(:name, :table) do
```
Comment (jcoleman) [inline lib/pg_ha_migrations/struct.rb:17]: I think Struct::Table and Struct::Index should maybe be the same thing. Or, at the very least, table should just be Relation, and Index should be a subclass of that.

File: lib/pg_ha_migrations/safe_statements.rb (hunk 7 above)
Comment (rkrage) [reply lib/pg_ha_migrations/safe_statements.rb:494]: Related to my comment above (https://github.com/braintree/pg_ha_migrations/pull/87#discussion_r1324759337), I was trying to draw a line with the struct objects such that they only do light validation / formatting and don't actually reach out to the database. If we want to cross that line, I think maybe we'd want to move some of the other helper methods as well like `partitioned?`, `child_partitions`, and `valid?`

File: lib/pg_ha_migrations/struct.rb (hunk 8 above)
Comment (rkrage) [reply lib/pg_ha_migrations/struct.rb]: Addressed in https://github.com/braintree/pg_ha_migrations/pull/87/commits/bb4b0ce47c6e88de1a0f96381583efa72a8b4495

File: lib/pg_ha_migrations/safe_statements.rb (hunk 7 above)
Comment (rkrage) [reply lib/pg_ha_migrations/safe_statements.rb]: Update: to hell with the line. This feels pretty clean: https://github.com/braintree/pg_ha_migrations/pull/87/commits/8c05071a751fc560ed02b730f011c22f155527f3

File: lib/pg_ha_migrations/struct.rb [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+
+      schema = connection.select_value(<<~SQL)
+        SELECT schemaname
+        FROM pg_tables
+        WHERE tablename = #{connection.quote(pg_name.identifier)} AND schemaname = #{schema_conditional}
+        ORDER BY array_position(current_schemas(false), schemaname)
+        LIMIT 1
+      SQL
+
+      raise ArgumentError, "Could not find table #{table}" unless schema.present?
```
Comment (rkrage) [inline lib/pg_ha_migrations/struct.rb:40]: This was previously a migration error, but it seems weird to raise that since this class can be used outside of migrations. Maybe we need to add a new error class?

File: lib/pg_ha_migrations/safe_statements.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (jcoleman) [reply lib/pg_ha_migrations/safe_statements.rb]: Ah, that's a good point. Take a string _or_ a table struct in would probably be good for a public function, but for the return value I think the struct instances would be good.

File: lib/pg_ha_migrations/safe_statements.rb [hunk 10]
```diff
@@ -582,16 +520,35 @@ def exec_migration(conn, direction)
...
+    nested_lock = Thread.current[__method__]
+
     _check_postgres_adapter!
-    table = table.to_s
-    quoted_table_name = connection.quote_table_name(table)
+
+    table = PgHaMigrations::Struct::Table.from_table_name(table)
+
+    # Disallow nested locks unless targeting the same table
+    if nested_lock && nested_lock != table
```
Comment (jcoleman) [inline lib/pg_ha_migrations/safe_statements.rb:530]: Now that this is the actual value we can simply:
```suggestion
    if nested_lock != table
```

File: spec/safe_statements_spec.rb [hunk 11]
```diff
@@ -3577,6 +3762,32 @@ def up
...
             end
 
+            it "raises error when attempting nested lock on different table" do
+              ActiveRecord::Base.connection.execute("CREATE TABLE foo(pk SERIAL, i INTEGER)")
+
+              expect do
+                migration.safely_acquire_lock_for_table(table_name) do
+                  expect(locks_for_table(table_name, connection: alternate_connection)).not_to be_empty
+
+                  # attempting a nested lock twice to ensure the
```
Comment (jcoleman) [inline spec/safe_statements_spec.rb:3772]: I like that this is tested.

File: lib/pg_ha_migrations/struct.rb (hunk 6 above)
Comment (jcoleman) [reply lib/pg_ha_migrations/struct.rb:15]: If we don't have a need for them now, that's fine, but I wouldn't shy from adding them if it made the code cleaner.

File: lib/pg_ha_migrations/safe_statements.rb (hunk 7 above)
Comment (jcoleman) [reply lib/pg_ha_migrations/safe_statements.rb]: Yeah, that seems a lot cleaner.

I wasn't going to mention it, but since you brought it up, I don't think the `Struct` nested module adds anything.

File: lib/pg_ha_migrations/safe_statements.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (rkrage) [reply lib/pg_ha_migrations/safe_statements.rb]: We could certainly have a public method that takes in a string and delegates to the struct:

```ruby
def partitions_for_table(table)
  PgHaMigrations::Struct::Table.from_table_name(table).partitions.map(&:fully_qualified_name)
end
```

File: lib/pg_ha_migrations/struct.rb (hunk 9 above)
Comment (jcoleman) [reply lib/pg_ha_migrations/struct.rb:40]: That's a good point. Probably still worth a PgHaMigrations scoped error class IMO.

File: lib/pg_ha_migrations/safe_statements.rb (hunk 10 above)
Comment (rkrage) [reply lib/pg_ha_migrations/safe_statements.rb:530]: I don't think so. In the case where there isn't a nested lock that would turn into `nil != table` which would evaluate to true and we'd incorrectly raise an error

File: lib/pg_ha_migrations/safe_statements.rb [hunk 12]
```diff
@@ -501,58 +482,15 @@ def _quoted_partman_schema
   end
 
   def _fully_qualified_table_name_for_partman(table)
```
Comment (jcoleman) [inline lib/pg_ha_migrations/safe_statements.rb:484]: I wouldn't worry about this now, but just to put this in the back of your head: this method is probably a candidate for moving to the Relation struct.

File: lib/pg_ha_migrations/struct.rb [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+
+        tables.concat(sub_partitions)
+      end
+
+      tables
+    end
+  end
+
+  class Index < Relation
+    # TODO: implement shortening to ensure < 63 bytes
```
Comment (jcoleman) [inline lib/pg_ha_migrations/struct.rb:83]: Can we add an Issue for this?

File: spec/safe_statements_spec.rb [hunk 14]
```diff
@@ -1,18 +1,26 @@
 require "spec_helper"
 
 RSpec.describe PgHaMigrations::SafeStatements do
-  TableLock = Struct.new(:table, :lock_type, :granted)
+  TableLock = Struct.new(:table, :lock_type, :granted, :pid)
   def locks_for_table(table, connection:)
+    identifiers = table.to_s.split(".")
```
Comment (jcoleman) [inline spec/safe_statements_spec.rb:6]: Will this ever come in with quoting?

File: spec/safe_statements_spec.rb [hunk 15]
```diff
@@ -1,18 +1,26 @@
...
+
     values = connection.execute(<<-SQL)
-      SELECT pg_class.relname AS table, pg_locks.mode AS lock_type, granted
+      SELECT pg_class.relname AS table, pg_locks.mode AS lock_type, granted, pid
       FROM pg_locks
       JOIN pg_class ON pg_locks.relation = pg_class.oid
+      JOIN pg_namespace ON pg_class.relnamespace = pg_namespace.oid
       WHERE pid IS DISTINCT FROM pg_backend_pid()
-        AND pg_class.relkind = 'r'
+        AND pg_class.relkind IN ('r', 'p')
```
Comment (jcoleman) [inline spec/safe_statements_spec.rb:18]:
```suggestion
        AND pg_class.relkind IN ('r', 'p') -- 'r' is a standard table; 'p' is a partition parent
```

File: spec/safe_statements_spec.rb [hunk 16]
```diff
@@ -1,18 +1,26 @@
...
-  TableLock = Struct.new(:table, :lock_type, :granted)
+  TableLock = Struct.new(:table, :lock_type, :granted, :pid)
   def locks_for_table(table, connection:)
+    identifiers = table.to_s.split(".")
+
+    identifiers.prepend("public") if identifiers.size == 1
+
+    schema, table = identifiers
+
     values = connection.execute(<<-SQL)
```
Comment (jcoleman) [inline spec/safe_statements_spec.rb:12]: I think we can now use this?
```suggestion
    values = connection.structs_from_sql(<<-SQL)
```

Note: this could totally be a separate PR.

File: spec/safe_statements_spec.rb [hunk 17]
```diff
@@ -3479,6 +3498,168 @@ def up
               end
             end
 
+            it "does not wait to acquire a lock if same table in different schema is blocked" do
```
Comment (jcoleman) [inline spec/safe_statements_spec.rb:3501]:
```suggestion
            it "does not wait to acquire a lock if a table with the same name but in different schema is blocked" do
```

File: spec/safe_statements_spec.rb (hunk 14 above)
Comment (rkrage) [reply spec/safe_statements_spec.rb:6]: I don't think so, but we could use `PgHaMigrations::Struct::Table.from_table_name(table)` if we wanted. This is just a test helper though so I don't feel too strongly

File: lib/pg_ha_migrations/safe_statements.rb (hunk 7 above)
Comment (rkrage) [reply lib/pg_ha_migrations/safe_statements.rb]: Addressed in https://github.com/braintree/pg_ha_migrations/pull/87/commits/bf4407d629c2b4eb8e5f1d23f85f9a8087e8fd06

File: lib/pg_ha_migrations/struct.rb (hunk 9 above)
Comment (rkrage) [reply lib/pg_ha_migrations/struct.rb]: Addressed in https://github.com/braintree/pg_ha_migrations/pull/87/commits/bf4407d629c2b4eb8e5f1d23f85f9a8087e8fd06

File: lib/pg_ha_migrations/safe_statements.rb [hunk 18]
```diff
@@ -582,16 +520,35 @@ def exec_migration(conn, direction)
...
+
     _check_postgres_adapter!
-    table = table.to_s
-    quoted_table_name = connection.quote_table_name(table)
+
+    table = PgHaMigrations::Struct::Table.from_table_name(table)
+
+    # Disallow nested locks unless targeting the same table
+    if nested_lock && nested_lock != table
+      raise PgHaMigrations::InvalidMigrationError, "Nested lock detected! Cannot acquire lock on #{table.fully_qualified_name} while #{nested_lock.fully_qualified_name} is locked."
```
[the remark this answers is not in the document]
Comment (rkrage) [reply lib/pg_ha_migrations/safe_statements.rb:531]: Addressed in https://github.com/braintree/pg_ha_migrations/pull/87/commits/bf4407d629c2b4eb8e5f1d23f85f9a8087e8fd06

File: spec/safe_statements_spec.rb (hunk 15 above)
Comment (rkrage) [reply spec/safe_statements_spec.rb]: Addressed in https://github.com/braintree/pg_ha_migrations/pull/87/commits/3d164e2cf3bced19fd3aca90fa360ede80d7c9db

File: spec/safe_statements_spec.rb (hunk 16 above)
Comment (rkrage) [reply spec/safe_statements_spec.rb]: Addressed in https://github.com/braintree/pg_ha_migrations/pull/87/commits/bf4407d629c2b4eb8e5f1d23f85f9a8087e8fd06

File: lib/pg_ha_migrations/safe_statements.rb (hunk 10 above)
Comment (jcoleman) [reply lib/pg_ha_migrations/safe_statements.rb:530]: Ah, I see that now.

**Review by jcoleman — APPROVED** [review]
Let's squash into relevant commits before merging.
